decision_service/include/factory_resolver.h

```diff
@@ -0,0 +1,23 @@
+#pragma once
+#include "model_mgmt.h"
+#include "object_factory.h"
+
+namespace reinforcement_learning  {
+  using dtfactory = utility::object_factory<model_management::i_data_transport>;
+  using modelfactory = utility::object_factory<model_management::i_model>;
+
+  extern dtfactory& data_transport_factory;
+  extern modelfactory& model_factory;
+
+  // For proper static intialization 
+  // Check https://en.wikibooks.org/wiki/More_C++_Idioms/Nifty_Counter for explanation
+  struct factory_initializer {
+    factory_initializer();
+    ~factory_initializer();
+    private:
+    static void register_default_factories();
+  };
+  // Every translation unit gets a factory_initializer
+  // only one translation unit will initialize it
+  static factory_initializer _init;
+}
```

Comment:
Consider passing in factories to rlclientlib.

Reply:
Will add a constructor that takes factories for the complex case.  The simple case will use the default factories.

---
In reply to: [196488055](https://github.com/JohnLangford/vowpal_wabbit/pull/1495#discussion_r196488055) [](ancestors = 196488055)